yt/data_objects/construction_data_containers.py

```diff
@@ -673,14 +676,26 @@ def get_data(self, fields = None):
                     "with nonzero num_ghost_zones." % self._num_ghost_zones)
             else:
                 raise
-        if len(part) > 0: self._fill_particles(part)
+
+        # checking if we have a sph particles
+        is_sph_field = False
+        if(hasattr(self.ds, '_sph_ptype') and len(part)>0):
+            is_sph_field = self.ds._sph_ptype in part[0]
+
+        if len(part) > 0 and len(alias) == 0:
+            if(is_sph_field):
+                self._fill_sph_particles(fields)
```

Comment:
I had to make the following change to avoid getting back particle data if I select the `temperature` field in a Gadget dataset (since it's a derived field, it's in `gen`, which means the data that gets filled in by `fill_sph_particles` gets clobbered below):

```
diff --git a/yt/data_objects/construction_data_containers.py b/yt/data_objects/construction_data_containers.py
index 757ae8aaa..3cac5cc58 100644
--- a/yt/data_objects/construction_data_containers.py
+++ b/yt/data_objects/construction_data_containers.py
@@ -685,6 +685,9 @@ class YTCoveringGrid(YTSelectionContainer3D):
         if len(part) > 0 and len(alias) == 0:
             if(is_sph_field):
                 self._fill_sph_particles(fields)
+                for field in fields:
+                    if field in gen:
+                        gen.remove(field)
             else:
                 self._fill_particles(part)
```